R/ui.R

```diff
@@ -103,17 +103,66 @@ ui_w <- function(text = "", env = parent.frame()) {
 #' # o Are you sure you did it right?
 #' }
 #' @noRd
-ui_throw <- function(message = "Internal error", details = character(0), env = parent.frame()) {
+ui_throw <- function(message = "Internal error", details = character(0),
+                     env = parent.frame(),
+                     glue_open = "{", glue_close = "}") {
   message <- cli_format_text(message, env = env)
 
-  if (length(details) != 0L) {
-    details <- glue::glue_collapse(details, sep = "\n")
-    message <- glue::glue(message, details, .sep = "\n")
-  }
+  error_messages <- purrr::map(
+    c(message, details),
+    glue,
+    .open = glue_open,
+    .close = glue_close
+  )
+
+  error_messages <- subset_lines_to_fit_in_limit(error_messages, 8000L)
+
+  message <- glue_collapse(error_messages, sep = "\n")
 
-  rlang::abort(message, class = "rextendr_error")
+  withr::with_options(
+    # Valid values are something between 1000 and 8170
+    # This will be set by {rlang} in the future release,
+    # https://github.com/r-lib/rlang/pull/1214
+    list(warning.length = 8000),
+    rlang::abort(message, class = "rextendr_error")
+  )
 }
 
 cli_format_text <- function(message, env = parent.frame()) {
   cli::cli_format_method(cli::cli_text(message, .envir = env))
 }
+
+#' Subset lines to fit within provided character limit.
+#'
+#' Verifies that `lines` fit within given limit of `max_length`.
+#' If not, lines are subset and an additional `truncation_notification`
+#' is appended to `lines`. The function ensures that the output message
+#' constructed by combining `lines` with `\n` separator
+#' fits within `max_length` limit.
+#'
+#' @param lines \[ ansi_character(n) \] Error messages.
+#' @param max_length Maximum total length of the collapsed message
+#' (measured using `nchar()`).
+#' @param truncation_notification Message pattern appended if any `lines`
+#' were removed. Understands `cli` interpolation and
+#' `n_removed` integer variable, which is set to the number of removed lines.
+#' @return \[ ansi_character(n) \] A subset of `lines` with
+#' `truncation_notification` appended (if needed).
+#' @noRd
+subset_lines_to_fit_in_limit <- function(lines,
+                                         max_length = 1000,
+                                         truncation_notification = "{.val {n_removed}} compiler messages not shown.") {
+  n_removed <- 100L
+  truncation_notification_size <- nchar(bullet_i(truncation_notification))
```

Comment:
By default, `nchar()` measures the numbers of the characters, but it seems it's actually measured in bytes (see the last one is trancated by 334 repeats).

```
> cat("\u2705")
✅> 
> nchar("\u2705")
[1] 1
> nchar("\u2705", type = "width")
[1] 2
> nchar("\u2705", type = "byte")
[1] 3
> 
> withr::with_options(
+   list(warning.length = 1000),
+   warning(stringr::str_dup("\u2705", 333))
+ )
Warning message:
In force(code) :
  ✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅
> 
> withr::with_options(
+   list(warning.length = 1000),
+   warning(stringr::str_dup("\u2705", 334))
+ )
Warning message:
In force(code) :
  ✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅ [... truncated]
```

Reply:
Interesting, the documentation on this topic is not as good as I expected, so I assumed we dealt with characters. I will use `type = "byte"`.